### PR TITLE
[FEAT] wault integration

### PR DESCRIPTION
**Pull request by almost-llamas**
## Description
New Farming Pool! Integrate Alpaca with Wault.
image

## ChangeLogs
- feat: Wault integration

### CODE REVIEW

File: test/Vault_WaultSwap.test.ts [hunk 1]
```diff
@@ -0,0 +1,2041 @@
...
+      // Alice liquidates Bob position#1
+      let aliceBefore = await baseToken.balanceOf(await alice.getAddress());
+      await expect(vaultAsAlice.kill('1')) // at health = 0.003000997994240237
+        .to.emit(vaultAsAlice, 'Kill')
+
+      let aliceAfter = await baseToken.balanceOf(await alice.getAddress());
+
+      // Bank balance is increase by liquidation (0.002700898194816214 = 0.9 * 0.003000997994240237)
+      AssertHelpers.assertAlmostEqual(
+        ethers.utils.parseEther('10.002700898194816214').toString(),
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:878]: assert with equation instead. Something like vaultBefore.add(0.0027xxx)

File: test/Vault_WaultSwap.test.ts [hunk 2]
```diff
@@ -0,0 +1,2041 @@
...
+      );
+
+      // ---------------- Reinvest#1 -------------------
+      // Wait for 1 day and someone calls reinvest
+      await TimeHelpers.increase(TimeHelpers.duration.days(ethers.BigNumber.from('1')));
+
+      let [workerLPBefore, workerDebtBefore] = await wexMaster.userInfo(poolId, waultSwapWorker.address);
+      await waultSwapWorkerAsEve.reinvest();
+
+      // PancakeWorker receives 303999999998816250 cake as a reward
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:1047]: should WaultswapWorker and WEX?

File: test/Vault_WaultSwap.test.ts [hunk 3]
```diff
@@ -0,0 +1,2041 @@
...
+      // Eve got 10% of 303999999998816250 cake = 0.01 * 303999999998816250 = 3039999999988162 bounty
+      AssertHelpers.assertAlmostEqual(
+        ethers.utils.parseEther('0.003039999999988162').toString(),
+        (await wex.balanceOf(await eve.getAddress())).toString(),
+      );
+
+      // Remaining PancakeWorker reward = 227999999998874730 - 22799999999887473 = 205199999998987257 (~90% reward)
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:1054]: WaultWorker + Add calculation

File: test/Vault_WaultSwap.test.ts [hunk 4]
```diff
@@ -0,0 +1,2041 @@
...
+      await waultSwapWorkerAsEve.reinvest();
+
+      // eve should earn wex as a reward for reinvest
+      AssertHelpers.assertAlmostEqual(
+        ethers.utils.parseEther('0.004559999999987660').toString(),
+        (await wex.balanceOf(await eve.getAddress())).toString(),
+      );
+
+      // Remaining Worker reward = 142858796296283038 - 14285879629628304 = 128572916666654734 (~90% reward)
+      // Convert 128572916666654734 uni to 157462478899282341 NATIVE
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:1094]: UNI -> WEX

File: test/Vault_WaultSwap.test.ts [hunk 5]
```diff
@@ -0,0 +1,2041 @@
...
+      // Convert NATIVE to 5001669421841640 LP token
+      [workerLPAfter, workerDebtAfter] = await wexMaster.userInfo(poolId, waultSwapWorker.address);
+      // LP tokens of worker should be inceased from reinvestment
+      expect(workerLPAfter).to.be.bignumber.gt(workerLPBefore);
+
+      // Check Bob position info
+      [bobHealth, bobDebtToShare] = await vault.positionInfo('1');
+      expect(bobHealth).to.be.bignumber.gt(ethers.utils.parseEther('20')); // Get Reward and increase health
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:1102]: if use greater, then use bobHealth.gt(bobHeathBefore)

File: test/Vault_WaultSwap.test.ts [hunk 6]
```diff
@@ -0,0 +1,2041 @@
...
+      AssertHelpers.assertAlmostEqual(
+        ethers.utils.parseEther('10').toString(),
+        bobDebtToShare.toString(),
+      );
+
+      // Check Alice position info
+      [aliceHealth, aliceDebtToShare] = await vault.positionInfo('2');
+      expect(aliceHealth).to.be.bignumber.gt(ethers.utils.parseEther('3')); // Get Reward and increase health
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:1110]: if use greater, then use aliceHealth.gt(aliceHealthBefore)

File: test/Vault_WaultSwap.test.ts [hunk 7]
```diff
@@ -0,0 +1,2041 @@
...
+      await waultSwapWorkerAsEve.reinvest();
+
+      // eve should earn wex as a reward for reinvest
+      AssertHelpers.assertAlmostEqual(
+        ethers.utils.parseEther('0.006079999999979926').toString(),
+        (await wex.balanceOf(await eve.getAddress())).toString(),
+      );
+
+      // Remaining Worker reward = 142858796296283038 - 14285879629628304 = 128572916666654734 (~90% reward)
+      // Convert 128572916666654734 uni to 74159218067697746 NATIVE
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:1130]: UNI -> WEX

File: test/Vault_WaultSwap.test.ts [hunk 8]
```diff
@@ -0,0 +1,2041 @@
...
+      // Convert NATIVE to 2350053120029788 LP token
+      [workerLPAfter, workerDebtAfter] = await wexMaster.userInfo(poolId, waultSwapWorker.address);
+      // LP tokens of worker should be inceased from reinvestment
+      expect(workerLPAfter).to.be.bignumber.gt(workerLPBefore);
+
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:1135]: assert Bob's health and Alice's health

File: test/Vault_WaultSwap.test.ts [hunk 9]
```diff
@@ -0,0 +1,2041 @@
...
+          ]
+        ),
+      );
+      const aliceAfter = await baseToken.balanceOf(await alice.getAddress());
+
+      // Check Alice account, Alice must be richer as she earned from leverage yield farm without getting liquidated
+      expect(aliceAfter).to.be.bignumber.gt(aliceBefore);
+    }).timeout(50000);
+
+    it('should liquidate user position correctly', async () => {
```
Comment (spicysquid168) [inline test/Vault_WaultSwap.test.ts:1196]: duplicated test-case?

File: test/WaultSwap_Restricted_StrategyAddBaseTokenOnly.test.ts [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+    // amountFTOKEN = 0.004649299362258152
+    // lpAmount = totalSupply * (amountF / reserveF) or totalSupply * (amountB / reserveB)
+    // lpAmount = 0.316242497512891045 * (0.004649299362258152 / 0.095350700637741848) ~= 0.015419981522648937...
+    // lpAmount = 0.316242497512891045 * (0.051142292984839684 / 1.048857707015160316) ~= 0.015419981522648941...
+
+    // actualLpAmount = 0.015419263215025115
+    expect(await lp.balanceOf(mockWaultSwapWorker.address)).to.be.bignumber.eq(ethers.utils.parseEther('0.015419263215025115'))
+    expect(await lp.balanceOf(strat.address)).to.be.bignumber.eq(ethers.utils.parseEther('0'))
+    expect(await farmingToken.balanceOf(strat.address)).to.be.bignumber.eq(ethers.utils.parseEther('0'))
+
```
Comment (spicysquid168) [inline test/WaultSwap_Restricted_StrategyAddBaseTokenOnly.test.ts:268]: assert baseToken.balanceOf(strat) = 0 too

File: test/WaultSwap_Restricted_StrategyAddTwoSidesOptimal.test.ts [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+      await baseToken.mint(mockWaultSwapWorker.address, ethers.utils.parseEther('2'))
+      await mockWaultSwapWorkerAsAlice.work(0, await alice.getAddress(), ethers.utils.parseEther('0'), ethers.utils.defaultAbiCoder.encode(
+        ['address', 'bytes'],
+        [addRestrictedStrat.address, ethers.utils.defaultAbiCoder.encode(
+          ['uint256','uint256'],
+          ['0', ethers.utils.parseEther('0.01')]
+        )],
+      ))
+
+      console.log((await baseToken.balanceOf(lpV2.address)).toString())
```
Comment (spicysquid168) [inline test/WaultSwap_Restricted_StrategyAddTwoSidesOptimal.test.ts:231]: remove console.log

File: test/WaultSwap_Restricted_StrategyLiquidate.test.ts [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+            [strat.address, ethers.utils.defaultAbiCoder.encode(
+            ['uint256'],
+            [ethers.utils.parseEther('1')]
+            )],
+        )
+    )
+
+    expect(await lp.balanceOf(strat.address)).to.be.bignumber.eq(ethers.utils.parseEther('0'));
+    expect(await lp.balanceOf(await bob.getAddress())).to.be.bignumber.eq(ethers.utils.parseEther('0'));
+    expect(await baseToken.balanceOf(lp.address)).to.be.bignumber.eq(ethers.utils.parseEther('0.500500500500500501'))
```
Comment (spicysquid168) [inline test/WaultSwap_Restricted_StrategyLiquidate.test.ts:263]: why 0.5005?

**Review by spicysquid168 — DISMISSED** [review]

**Review by spicysquid168 — APPROVED** [review]

**Review by abraham-alpaca — APPROVED** [review]

**Pull request merged**